src/array.rs

```diff
@@ -0,0 +1,176 @@
+//! Extends non-zero length arrays with conversion methods to non-empty
+//! collections.
+//!
+//! Since fixed-size arrays are by definition already not empty, they aren't
+//! given a special wrapper type like [`crate::NEVec`]. Instead, we enable them
+//! to be easily iterated over in a compatible way:
+//!
+//! ```
+//! use nonempty_collections::*;
+//!
+//! let a: [u32; 4] = [1, 2, 3, 4];
+//! let v: NEVec<_> = a.into_nonempty_iter().map(|n| n + 1).collect();
+//! assert_eq!(nev![2, 3, 4, 5], v);
+//! ```
+//!
+//! See [`NonEmptyArrayExt`] for more conversions.
+
+use std::num::NonZeroUsize;
+
+use crate::impl_nonempty_iter_for_arrays;
+use crate::IntoNonEmptyIterator;
+use crate::NonEmptyIterator;
+
+/// Provides extension methods for non-empty arrays.
+///
+/// # Examples
+///
+/// Create a non-empty slice of an array:
+///
+/// ```
+/// # use nonempty_collections::*;
+/// assert_eq!(NESlice::new(&1, &[2]), [1, 2].as_nonempty_slice());
+/// ```
+///
+/// Get the length of an array as a [`NonZeroUsize`]:
+///
+/// ```
+/// # use nonempty_collections::NonEmptyArrayExt;
+/// # use std::num::NonZeroUsize;
+/// assert_eq!(NonZeroUsize::MIN, [1].nonzero_len());
+/// ```
+///
+/// Copy into a non-empty vec:
+///
+/// ```
+/// # use nonempty_collections::*;
+/// assert_eq!(nev![4], [4].to_nonempty_vec());
+/// ```
+pub trait NonEmptyArrayExt<T> {
+    /// Create a `NESlice` that borrows the contents of `self`.
+    fn as_nonempty_slice(&self) -> crate::NESlice<'_, T>;
+
+    /// Returns the length of this array as a [`NonZeroUsize`].
+    fn nonzero_len(&self) -> NonZeroUsize;
+
+    /// Copies `self` into a new [`crate::NEVec`].
+    fn to_nonempty_vec(&self) -> crate::NEVec<T>
+    where
+        T: Clone;
```

Comment:
I suspect this should be `into_nonempty_vec`, we're able to do ownership passing of values in an array via `into_iter`:

https://doc.rust-lang.org/std/primitive.array.html#impl-IntoIterator-for-%5BT;+N%5D

This would allow us to avoid the `Clone`. If the user doesn't want to transfer ownership, they're free to call `as_nonempty_slice`.

Reply:
It was actually meant to mimic [`slice::to_vec()`](https://doc.rust-lang.org/std/primitive.slice.html#method.to_vec) but as I'm writing this I realize that this is available on arrays due to `Deref` and not directly implemented on array.

I converted the method into `into_nonempty_vec` as you suggested 👍